File: bindings/python/example_stcp_visitor.py
#!/usr/bin/env python3
"""
Example: STCP Visitor using Python FRPC bindings
This example acts as a client (visitor) that connects to a remote STCP service.
"""

import sys
import time
import socket
import threading
from frpc_python import FRPCClient, TunnelType, LogLevel, set_log_callback, cleanup

def log_handler(level, message):
    """Handle log messages from FRPC"""
    level_names = {
        LogLevel.DEBUG: "DEBUG",
        LogLevel.INFO: "INFO", 
        LogLevel.WARN: "WARN",
        LogLevel.ERROR: "ERROR"
    }
    print(f"[{level_names.get(level, 'UNKNOWN')}] {message}")

def on_data_received(data):
    """Handle data received from tunnel"""
    print(f"Visitor received data: {data.decode('utf-8', errors='ignore')}")

def on_connection_changed(connected, error_code):
    """Handle connection status changes"""
    if connected:
        print("Visitor tunnel connected successfully")
    else:
        print(f"Visitor tunnel disconnected (error: {error_code})")

def main():
    # Configuration
    server_addr = "127.0.0.1"
    server_port = 7000
    token = None  # No token for testing; set actual token if frps requires authentication
    
    # STCP Visitor Config
    tunnel_name = "python_stcp_visitor"
    server_name = "python_stcp_server" # Must match the server's tunnel name
    secret_key = "python_secret"       # Must match the server's secret
    bind_addr = "127.0.0.1"
    bind_port = 9090
    
    print("Python FRPC STCP Visitor Example")
    print(f"Connecting to FRP server: {server_addr}:{server_port}")
    print(f"Target Tunnel: {server_name}")
    print(f"Listening on: {bind_addr}:{bind_port}")
    print("Press Ctrl+C to exit\n")
    
    # Set up logging
    set_log_callback(log_handler)
    
    client = None
    tunnel = None
    
    try:
        # Create FRPC client
        client = FRPCClient(server_addr, server_port, token)
        
        # Create STCP visitor tunnel
        # Note: 'data_callback' is optional for visitors. 
        # If you want to intercept data transparently, you can use it,
        # but normally visitors just open a local port (bind_port).
        tunnel = client.create_tunnel(
            TunnelType.STCP_VISITOR,
            tunnel_name,
            secret_key=secret_key,
            remote_name=server_name,
            bind_addr=bind_addr,
            bind_port=bind_port,
            connection_callback=on_connection_changed
        )
        
        # Connect to FRP server
        print("Connecting to FRP server...")
        client.connect()
        
        # Start the tunnel
        print("Starting STCP visitor tunnel...")
        tunnel.start()
        
        print(f"Visitor is running. You can now connect to {bind_addr}:{bind_port}")
        
        # Keep the program running and show stats periodically
        while True:
            time.sleep(5)
            if client.is_connected() and tunnel.is_active():
                stats = tunnel.get_stats()
                print(f"Stats - Sent: {stats['bytes_sent']} bytes, "
                      f"Received: {stats['bytes_received']} bytes, "
                      f"Active connections: {stats['connections_active']}")
            else:
                print("Warning: Client disconnected or tunnel inactive")
                
    except KeyboardInterrupt:
        print("\nShutting down...")
    except Exception as e:
        print(f"Error: {e}")
    finally:
        # Cleanup
        if tunnel:
            tunnel.close()
        if client:
            client.close()
        cleanup()
        print("Cleanup completed")

if __name__ == "__main__":
    main()

File: release/bindings/nodejs/example_stcp_visitor.js
/**
 * Example: STCP Visitor using Node.js FRPC bindings
 * This example creates an STCP visitor that connects to a remote STCP server.
 */

const { FRPCClient, TunnelType, cleanup } = require('./frpc_node');
const net = require('net');

// Configuration
const config = {
    serverAddr: '127.0.0.1',
    serverPort: 7000,
    token: 'test_token',
    tunnelName: 'nodejs_stcp_visitor',
    secretKey: 'nodejs_secret',
    remoteName: 'python_stcp_server', // Connect to the Python server example
    bindAddr: '127.0.0.1',
    bindPort: 9090
};

console.log('Node.js FRPC STCP Visitor Example');
console.log(`Connecting to FRP server: ${config.serverAddr}:${config.serverPort}`);
console.log(`Tunnel: ${config.tunnelName}`);
console.log(`Remote server: ${config.remoteName}`);
console.log(`Local bind: ${config.bindAddr}:${config.bindPort}`);
console.log('Press Ctrl+C to exit\n');

let client = null;
let tunnel = null;
let localServer = null;

async function startLocalServer() {
    return new Promise((resolve, reject) => {
        localServer = net.createServer((socket) => {
            console.log(`Local client connected: ${socket.remoteAddress}:${socket.remotePort}`);
            
            socket.on('data', (data) => {
                console.log(`Forwarding data to tunnel: ${data.toString()}`);
                try {
                    tunnel.sendData(data);
                } catch (error) {
                    console.error('Error sending data through tunnel:', error);
                }
            });
            
            socket.on('close', () => {
                console.log('Local client disconnected');
            });
            
            socket.on('error', (error) => {
                console.error('Local socket error:', error);
            });
        });
        
        localServer.listen(config.bindPort, config.bindAddr, () => {
            console.log(`Local server listening on ${config.bindAddr}:${config.bindPort}`);
            resolve();
        });
        
        localServer.on('error', reject);
    });
}

async function main() {
    try {
        // Start local server first
        await startLocalServer();
        
        // Create FRPC client
        client = new FRPCClient(config.serverAddr, config.serverPort, config.token);
        
        // Create STCP visitor tunnel
        tunnel = client.createTunnel(TunnelType.STCP_VISITOR, config.tunnelName, {
            secretKey: config.secretKey,
            remoteName: config.remoteName,
            bindAddr: config.bindAddr,
            bindPort: config.bindPort,
            onData: (data) => {
                console.log(`Received data from tunnel: ${data.toString()}`);
                // In a real implementation, you'd forward this to the appropriate local connection
            },
            onConnection: (connected, errorCode) => {
                if (connected) {
                    console.log('Visitor tunnel connected successfully');
                } else {
                    console.log(`Visitor tunnel disconnected (error: ${errorCode})`);
                }
            }
        });
        
        // Set up event handlers
        client.on('connected', () => {
            console.log('Client connected to FRP server');
            tunnel.start();
        });
        
        client.on('disconnected', () => {
            console.log('Client disconnected from FRP server');
        });
        
        client.on('error', (error) => {
            console.error('Client error:', error);
        });
        
        tunnel.on('started', () => {
            console.log('STCP visitor tunnel started');
            console.log('You can now connect to the local server to test the tunnel');
        });
        
        tunnel.on('error', (error) => {
            console.error('Tunnel error:', error);
        });
        
        // Connect to FRP server
        console.log('Connecting to FRP server...');
        client.connect();
        
        // Show stats periodically
        const statsInterval = setInterval(() => {
            if (client.isConnected() && tunnel.isActive()) {
                const stats = tunnel.getStats();
                console.log(`Stats - Sent: ${stats.bytesSent} bytes, ` +
                           `Received: ${stats.bytesReceived} bytes, ` +
                           `Active connections: ${stats.connectionsActive}`);
            } else {
                console.log('Warning: Client disconnected or tunnel inactive');
            }
        }, 10000); // Every 10 seconds
        
        // Graceful shutdown
        process.on('SIGINT', () => {
            console.log('\nShutting down...');
            clearInterval(statsInterval);
            
            if (localServer) {
                localServer.close();
            }
            if (client) {
                client.close();
            }
            cleanup();
            
            console.log('Cleanup completed');
            process.exit(0);
        });
        
        // Keep the process running
        console.log('STCP visitor is running. Waiting for connections...');
        
    } catch (error) {
        console.error('Error:', error);
        
        // Cleanup on error
        if (localServer) {
            localServer.close();
        }
        if (client) {
            client.close();
        }
        cleanup();
        
        process.exit(1);
    }
}

// Test function to send data through the tunnel
function testTunnel() {
    if (!tunnel || !tunnel.isActive()) {
        console.log('Tunnel not active, cannot send test data');
        return;
    }
    
    const testData = `Hello from Node.js visitor at ${new Date().toISOString()}`;
    console.log(`Sending test data: ${testData}`);
    
    try {
        tunnel.sendData(testData);
    } catch (error) {
        console.error('Error sending test data:', error);
    }
}

// Export test function for external use
module.exports = { testTunnel };

// Run if this is the main module
if (require.main === module) {
    main();
}